domain/src/test/java/org/oppia/domain/OnboardingFlowControllerTest.kt

```diff
@@ -0,0 +1,247 @@
+  fun testControllerObserver_observedAfterSettingAppOnboarded_providesLiveData_userDidNotOnboardApp() =
+    runBlockingTest(coroutineContext) {
+      val onboardingg = onboardingFlowController.getOnboardingFlow()
+
+      onboardingg.observeForever(mockOnboardingObserver)
+      onboardingFlowController.markOnboardingFlowCompleted()
+      advanceUntilIdle()
+
+      // The result should not indicate that the user onboarded the app because markUserOnboardedApp does not notify observers
+      // of the change.
+      verify(mockOnboardingObserver, atLeastOnce()).onChanged(onboardinggResultCaptor.capture())
+      assertThat(onboardinggResultCaptor.value.isSuccess()).isTrue()
+      assertThat(onboardinggResultCaptor.value.getOrThrow().alreadyOnboardedApp).isFalse()
+    }
+
+  @Test
+  @ExperimentalCoroutinesApi
+  fun testController_settingAppOnboarded_observedNewController_userOnboardedApp() = runBlockingTest(coroutineContext) {
+    onboardingFlowController.markOnboardingFlowCompleted()
+    advanceUntilIdle()
+
+    // Create the controller by creating another singleton graph and injecting it (simulating the app being recreated).
+    setUpTestApplicationComponent()
+    val onboardingg = onboardingFlowController.getOnboardingFlow()
+    onboardingg.observeForever(mockOnboardingObserver)
+    advanceUntilIdle()
+
+    // The app should be considered onboarded since a new LiveData instance was observed after marking the app as onboarded.
+    verify(mockOnboardingObserver, atLeastOnce()).onChanged(onboardinggResultCaptor.capture())
+    assertThat(onboardinggResultCaptor.value.isSuccess()).isTrue()
+    assertThat(onboardinggResultCaptor.value.getOrThrow().alreadyOnboardedApp).isTrue()
+  }
+
+  @Test
+  @ExperimentalCoroutinesApi
+  fun testController_onboardedApp_cleared_observeNewController_userDidNotOnboardApp() =
+    runBlockingTest(coroutineContext) {
+      val onboardingFlowStore = cacheFactory.create("on_boarding_flow", OnboardingFlow.getDefaultInstance())
+      onboardingFlowController.markOnboardingFlowCompleted()
+      advanceUntilIdle()
+      clearOnboardingFlow(onboardingFlowStore)
+      setUpTestApplicationComponent()
+      val onboardingg = onboardingFlowController.getOnboardingFlow()
+      onboardingg.observeForever(mockOnboardingObserver)
+      advanceUntilIdle()
+      // The app should be considered not yet onboarded since the previous history was cleared.
+      verify(mockOnboardingObserver, atLeastOnce()).onChanged(onboardinggResultCaptor.capture())
+      assertThat(onboardinggResultCaptor.value.isSuccess()).isTrue()
+      assertThat(onboardinggResultCaptor.value.getOrThrow().alreadyOnboardedApp).isFalse()
+    }
+
+  /** Clears any indication that the user has previously completed onboarding the application. */
+  private fun clearOnboardingFlow(onboardingFlowStore: PersistentCacheStore<OnboardingFlow>) {
+    onboardingFlowStore.clearCacheAsync().invokeOnCompletion {
+      it?.let {
```

Comment:
This can actually be removed as we are not anything inside this.
`onboardingFlowStore.clearCacheAsync()` this is enough.

Reply:
Done